<a href="https://colab.research.google.com/github/NavjyotDataScientist/go_for_small_project_first_then_for_world/blob/main/heart_failure_clinical_records.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Phase 1:

-Data Understaing : How to check our dataset from different angles and simply says just observe and write down the key points what things we need to change as per the company requirements.

- Make sure you understand each small step to tackle the complex structure or problem statement.


In [1]:
import pandas as pd

# Load the dataset
df = pd.read_csv('/content/drive/MyDrive/heart_failure_clinical_records/heart_failure_clinical_records_dataset.csv')

print('Dataset loaded successfully.')

Dataset loaded successfully.


In [2]:
df.index

RangeIndex(start=0, stop=299, step=1)

Observation:

- As we can see the above we have 0 means it start from 0 and 299 means we have rows and step 1 means it works like 1, 2, 3, 4..

In [3]:
df.columns

Index(['age', 'anaemia', 'creatinine_phosphokinase', 'diabetes',
       'ejection_fraction', 'high_blood_pressure', 'platelets',
       'serum_creatinine', 'serum_sodium', 'sex', 'smoking', 'time',
       'DEATH_EVENT'],
      dtype='object')

Observation:

- check each columns names as per the best practice make it in lower case for better readability and make sense for understanding for model which will going to predict.

- lets say DEATH_EVENT change this in cleaning file and write a note about that.
- this way we have a undertanding the data here and make changes in cleaning step. I hope everyone will got this captain.

In [4]:
df.dtypes

age                         float64
anaemia                       int64
creatinine_phosphokinase      int64
diabetes                      int64
ejection_fraction             int64
high_blood_pressure           int64
platelets                   float64
serum_creatinine            float64
serum_sodium                  int64
sex                           int64
smoking                       int64
time                          int64
DEATH_EVENT                   int64
dtype: object

observation:

- So again my dear we need to carefully write and understand here and whats that see here below and above.
- age columns we have in float64 so to understand this we cant write all the things at this file for this Example - the age is 23, or 45 so we dont need 23.0 or lets say 65.0 or 56.0 we want only simple like 23, 65, 78 i hope you got that.
- make it in int16 which is risk free and memory efficient mostly i am right here. you have to get the things of little bit knowledge to build a good model we need a error free dataset. So for that reason we are wasting time by writing observation in each cell.
-

### First 5 rows of the dataset:

In [5]:
display(df.head())

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0  75.0        0                       582         0                 20   
1  55.0        0                      7861         0                 38   
2  65.0        0                       146         0                 20   
3  50.0        1                       111         0                 20   
4  65.0        1                       160         1                 20   

   high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                    1  265000.00               1.9           130    1   
1                    0  263358.03               1.1           136    1   
2                    0  162000.00               1.3           129    1   
3                    0  210000.00               1.9           137    1   
4                    0  327000.00               2.7           116    0   

   smoking  time  DEATH_EVENT  
0        0     4            1  
1        0     6            1  
2        1     7            1  
3        0     7            1  
4        0     8            1

observation:

- by seeing the dataset what actually we see means just observe and write down all the points.
- make the death_event column in lower case first for better readability and make sense too.
- if something we missed dont panic lets say any key point or observation write it down on another cell.



In [6]:
display(df.tail())

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
294  62.0        0                        61         1                 38   
295  55.0        0                      1820         0                 38   
296  45.0        0                      2060         1                 60   
297  45.0        0                      2413         0                 38   
298  50.0        0                       196         0                 45   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
294                    1   155000.0               1.1           143    1   
295                    0   270000.0               1.2           139    0   
296                    0   742000.0               0.8           138    0   
297                    0   140000.0               1.4           140    1   
298                    0   395000.0               1.6           136    1   

     smoking  time  DEATH_EVENT  
294        1   270            0  
295        0   271            0  
296        0   278            0  
297        1   280            0  
298        1   285            0

Observation:

-Its name suggest we can see the last 5 rows of our dataset simply check the index number for confirmation.


### Dataset Information (Column types, non-null values):

In [7]:
display(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 13 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   age                       299 non-null    float64
 1   anaemia                   299 non-null    int64  
 2   creatinine_phosphokinase  299 non-null    int64  
 3   diabetes                  299 non-null    int64  
 4   ejection_fraction         299 non-null    int64  
 5   high_blood_pressure       299 non-null    int64  
 6   platelets                 299 non-null    float64
 7   serum_creatinine          299 non-null    float64
 8   serum_sodium              299 non-null    int64  
 9   sex                       299 non-null    int64  
 10  smoking                   299 non-null    int64  
 11  time                      299 non-null    int64  
 12  DEATH_EVENT               299 non-null    int64  
dtypes: float64(3), int64(10)
memory usage: 30.5 KB


None

### Observations from `df.info()`:

*   **Total Entries**: The dataset contains 299 entries (rows), indexed from 0 to 298.
*   **Total Columns**: There are 13 columns in total.
*   **No Missing Values**: All columns have 299 non-null values, indicating that there are no missing data points in any of the features.
*   **Data Types**:
    *   **Numerical (Float64)**: `age`, `platelets`, `serum_creatinine`
    *   **Numerical (Int64)**: `anaemia`, `creatinine_phosphokinase`, `diabetes`, `ejection_fraction`, `high_blood_pressure`, `serum_sodium`, `sex`, `smoking`, `time`, `DEATH_EVENT`
*   **Memory Usage**: The dataset consumes approximately 30.5 KB of memory.

These observations provide a solid foundation for understanding the completeness and basic structure of our data before proceeding with further analysis.

Observation:

- the above info is from the ai tools where we get help if we want any and use it wisely.
- lets say see --- aware --- alerts---decide what to do---make changes-- accordingly.


In [8]:
df.values

array([[7.500e+01, 0.000e+00, 5.820e+02, ..., 0.000e+00, 4.000e+00,
        1.000e+00],
       [5.500e+01, 0.000e+00, 7.861e+03, ..., 0.000e+00, 6.000e+00,
        1.000e+00],
       [6.500e+01, 0.000e+00, 1.460e+02, ..., 1.000e+00, 7.000e+00,
        1.000e+00],
       ...,
       [4.500e+01, 0.000e+00, 2.060e+03, ..., 0.000e+00, 2.780e+02,
        0.000e+00],
       [4.500e+01, 0.000e+00, 2.413e+03, ..., 1.000e+00, 2.800e+02,
        0.000e+00],
       [5.000e+01, 0.000e+00, 1.960e+02, ..., 1.000e+00, 2.850e+02,
        0.000e+00]])

In [9]:
df.axes

[RangeIndex(start=0, stop=299, step=1),
 Index(['age', 'anaemia', 'creatinine_phosphokinase', 'diabetes',
        'ejection_fraction', 'high_blood_pressure', 'platelets',
        'serum_creatinine', 'serum_sodium', 'sex', 'smoking', 'time',
        'DEATH_EVENT'],
       dtype='object')]

In [10]:
df.ndim

2

In [11]:
df.size

3887

In [12]:
df.shape

(299, 13)

Observation:

- It shows number of rows and column

In [17]:
df.memory_usage(deep=True)

Index                        132
age                         2392
anaemia                     2392
creatinine_phosphokinase    2392
diabetes                    2392
ejection_fraction           2392
high_blood_pressure         2392
platelets                   2392
serum_creatinine            2392
serum_sodium                2392
sex                         2392
smoking                     2392
time                        2392
DEATH_EVENT                 2392
dtype: int64

In [19]:
df.empty

False

Observation:

- So we have plethora ways to see our dataset in different perspective.

- It will also seems that do not make sense However when you come accross on any project and you need the detail info for that reason we have all the code which I have been using here.

### Descriptive Statistics:

In [20]:
display(df.describe())

age     anaemia  creatinine_phosphokinase    diabetes  \
count  299.000000  299.000000                299.000000  299.000000   
mean    60.833893    0.431438                581.839465    0.418060   
std     11.894809    0.496107                970.287881    0.494067   
min     40.000000    0.000000                 23.000000    0.000000   
25%     51.000000    0.000000                116.500000    0.000000   
50%     60.000000    0.000000                250.000000    0.000000   
75%     70.000000    1.000000                582.000000    1.000000   
max     95.000000    1.000000               7861.000000    1.000000   

       ejection_fraction  high_blood_pressure      platelets  \
count         299.000000           299.000000     299.000000   
mean           38.083612             0.351171  263358.029264   
std            11.834841             0.478136   97804.236869   
min            14.000000             0.000000   25100.000000   
25%            30.000000             0.000000  212500.000000   
50%            38.000000             0.000000  262000.000000   
75%            45.000000             1.000000  303500.000000   
max            80.000000             1.000000  850000.000000   

       serum_creatinine  serum_sodium         sex    smoking        time  \
count         299.00000    299.000000  299.000000  299.00000  299.000000   
mean            1.39388    136.625418    0.648829    0.32107  130.260870   
std             1.03451      4.412477    0.478136    0.46767   77.614208   
min             0.50000    113.000000    0.000000    0.00000    4.000000   
25%             0.90000    134.000000    0.000000    0.00000   73.000000   
50%             1.10000    137.000000    1.000000    0.00000  115.000000   
75%             1.40000    140.000000    1.000000    1.00000  203.000000   
max             9.40000    148.000000    1.000000    1.00000  285.000000   

       DEATH_EVENT  
count    299.00000  
mean       0.32107  
std        0.46767  
min        0.00000  
25%        0.00000  
50%        0.00000  
75%        1.00000  
max        1.00000

### Observation:

- The `describe()` method provides a statistical summary of the numerical columns. We can observe the count, mean, standard deviation, min, max, and quartile values for each numerical feature. This helps in understanding the spread and central tendency of the data. For instance, 'age' ranges from 40 to 95, and 'creatinine_phosphokinase' has a very wide range, indicating potential outliers or a skewed distribution.
- The 'DEATH_EVENT' column has a mean of approximately 0.32, suggesting about 32% of the patients in the dataset experienced a death event.

### Missing Values Check:

In [21]:
display(df.isnull().sum())

age                         0
anaemia                     0
creatinine_phosphokinase    0
diabetes                    0
ejection_fraction           0
high_blood_pressure         0
platelets                   0
serum_creatinine            0
serum_sodium                0
sex                         0
smoking                     0
time                        0
DEATH_EVENT                 0
dtype: int64

### Observation:

- As confirmed by `df.info()`, `df.isnull().sum()` explicitly shows that there are no missing values across any of the columns. This means the dataset is complete, and we won't need to perform any imputation for missing data.

### Distribution of Target Variable ('DEATH_EVENT'):

In [22]:
display(df['DEATH_EVENT'].value_counts())
display(df['DEATH_EVENT'].value_counts(normalize=True))

DEATH_EVENT
0    203
1     96
Name: count, dtype: int64

DEATH_EVENT
0    0.67893
1    0.32107
Name: proportion, dtype: float64

### Observation:

- The target variable `DEATH_EVENT` shows that there are 203 instances where the event did not occur (0) and 96 instances where it did occur (1).
- In terms of percentages, approximately 67.89% of the patients did not experience a death event, while 32.11% did. This indicates a class imbalance, which is important to consider during model training.